### PR TITLE
[SPARK-22465][Core] Add a safety-check to RDD defaultPartitioner

### PR DESCRIPTION
**Pull request by sujithjay**
## What changes were proposed in this pull request?
In choosing a Partitioner to use for a cogroup-like operation between a number of RDDs, the default behaviour was if some of the RDDs already have a partitioner, we choose the one amongst them with the maximum number of partitions.

This behaviour, in some cases, could hit the 2G limit (SPARK-6235). To illustrate one such scenario, consider two RDDs:
rDD1: with smaller data and smaller number of partitions, alongwith a Partitioner.
rDD2: with much larger data and a larger number of partitions, without a Partitioner.

The cogroup of these two RDDs could hit the 2G limit, as a larger amount of data is shuffled into a smaller number of partitions.

This PR introduces a safety-check wherein the Partitioner is chosen only if either of the following conditions are met:
1. if the number of partitions of the RDD associated with the Partitioner is greater than or equal to the max number of upstream partitions; or 
2. if the number of partitions of the RDD associated with the Partitioner is less than and within a single order of magnitude of the max number of upstream partitions.

## How was this patch tested?
Unit tests in PartitioningSuite and PairRDDFunctionsSuite

### CODE REVIEW

**Comment by sujithjay** [comment]
SparkR test failure seems unrelated to this PR. Any ideas what's wrong?

**Comment by sujithjay** [comment]
SparkR test failure seems unrelated to this PR. Any ideas what's wrong?

**Comment by sujithjay** [comment]
Hi @HyukjinKwon , can you please help me with these SparkR tests failures? They seem unrelated to me.

**Comment by sujithjay** [comment]
cc: @tgravescs @codlife Could you please review this?

**Comment by HyukjinKwon** [comment]
Yup, AppVeyor test seems unrelated. In a quick look, AppVeyor is related with the latest `testthat` (1.0.2 -> 2.0.0). Will take a look for this separately.

**Comment by HyukjinKwon** [comment]
It was fixed in https://github.com/apache/spark/pull/20003. Rebasing should make the test apssed.

**Comment by sujithjay** [comment]
Thank you, @HyukjinKwon . The tests passed after rebasing.

**Comment by sujithjay** [comment]
@tgravescs , could you please take a look when you have some time ?

**Comment by tgravescs** [comment]
Jenkins, test this please

**Comment by tgravescs** [comment]
ok to test

**Comment by tgravescs** [comment]
@sujithjay thanks for working on this.  I will review but I'm not sure I will get to it for a bit, I'm out for the holidays and not sure I can give this the time it needs for a full review today.

**Comment by sujithjay** [comment]
@tgravescs Thank you for keeping me informed. I look forward to receiving your review. Happy holidays!

File: core/src/main/scala/org/apache/spark/Partitioner.scala [hunk 1]
```diff
@@ -57,7 +60,8 @@ object Partitioner {
   def defaultPartitioner(rdd: RDD[_], others: RDD[_]*): Partitioner = {
     val rdds = (Seq(rdd) ++ others)
     val hasPartitioner = rdds.filter(_.partitioner.exists(_.numPartitions > 0))
-    if (hasPartitioner.nonEmpty) {
+    if (hasPartitioner.nonEmpty
+      && isEligiblePartitioner(hasPartitioner.maxBy(_.partitions.length), rdds)) {
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: `hasPartitioner.maxBy(_.partitions.length)` is used repeatedly, pull that into a variable ?

File: core/src/main/scala/org/apache/spark/Partitioner.scala [hunk 2]
```diff
@@ -67,6 +71,16 @@ object Partitioner {
...
   }
+
+  /**
+   * Returns true if the number of partitions of the RDD is either greater than or is
+   * less than and within a single order of magnitude of the max number of upstream partitions;
+   * otherwise, returns false
+   */
+  private def isEligiblePartitioner(hasMaxPartitioner: RDD[_], rdds: Seq[RDD[_]]): Boolean = {
+    val maxPartitions = rdds.map(_.partitions.length).max
+    log10(maxPartitions).floor - log10(hasMaxPartitioner.getNumPartitions).floor < 1
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: Why `.floor` ?
It causes unnecessary discontinuity imo, for example: (9, 11) will not satisfy - but it should.

Comment (sujithjay) [reply core/src/main/scala/org/apache/spark/Partitioner.scala]: Hi @mridulm , I suppose I was trying to ensure a strict order-of-magnitude check; but, I agree it leads to a discontinuity. I will change this, and the corresponding test cases.

**Comment by sujithjay** [comment]
Scala style tests are failing on a file 'SparkHiveExample.scala' , which is unrelated to this PR. Will rebase to master and try again.

**Comment by HyukjinKwon** [comment]
@sujithjay, I opened a hotfix. It should be fine soon (maybe after few hours).

**Comment by sujithjay** [comment]
Thank you, @HyukjinKwon . I will try again after the hotfix is merged to master.

File: core/src/main/scala/org/apache/spark/Partitioner.scala [hunk 3]
```diff
@@ -21,6 +21,8 @@ import java.io.{IOException, ObjectInputStream, ObjectOutputStream}
 
 import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
+import scala.language.existentials
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: Curious, why was this required ?

Comment (sujithjay) [reply core/src/main/scala/org/apache/spark/Partitioner.scala]: Without this import, there was a compiler warning:
```
Warning:(63, 29) inferred existential type Option[org.apache.spark.rdd.RDD[_$2]]( forSome { type _$2 }), which cannot be expressed by wildcards,  should be enabled
by making the implicit value scala.language.existentials visible.
This can be achieved by adding the import clause 'import scala.language.existentials'
or by setting the compiler option -language:existentials.
See the Scaladoc for value scala.language.existentials for a discussion
why the feature should be explicitly enabled.
```

Spark build failed because of this.

Comment (mridulm) [reply core/src/main/scala/org/apache/spark/Partitioner.scala]: If we explicitly set the type, is it still required ? For example, with `val hasMaxPartitioner: Option[RDD[_]] = ...` ?

File: core/src/test/scala/org/apache/spark/PartitioningSuite.scala [hunk 4]
```diff
@@ -259,6 +259,27 @@ class PartitioningSuite extends SparkFunSuite with SharedSparkContext with Priva
...
+
+    val partitioner1 = Partitioner.defaultPartitioner(rdd1, rdd2)
+    val partitioner2 = Partitioner.defaultPartitioner(rdd2, rdd3)
+    val partitioner3 = Partitioner.defaultPartitioner(rdd3, rdd1)
+    val partitioner4 = Partitioner.defaultPartitioner(rdd1, rdd2, rdd3)
+
+    assert(partitioner1.numPartitions == rdd1.getNumPartitions)
+    assert(partitioner2.numPartitions == rdd3.getNumPartitions)
+    assert(partitioner3.numPartitions == rdd3.getNumPartitions)
+    assert(partitioner4.numPartitions == rdd3.getNumPartitions)
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/PartitioningSuite.scala]: Can you add a testcase such that numPartitions 9 vs 11 is not treated as an order of magnitude jump (to prevent future changes which end up breaking this).

**Comment by mridulm** [comment]
I left a couple of comments @sujithjay, overall it is looking good, thanks for working on it !
We can merge it once they are addressed.

**Comment by sujithjay** [comment]
Thank you, @mridulm for reviewing this PR.  I have addressed the latest review comments.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by mridulm** [comment]
Looks good @sujithjay ... once we have a successful build, I will merge it in.

**Comment by sujithjay** [comment]
The failed unit test (in HistoryServerSuite.scala) seems unrelated to this PR.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by mridulm** [comment]
Merged, thanks for fixing this @sujithjay !
